content/blog/2019-10-3-tdd-gimmick/index.md

```diff
@@ -0,0 +1,196 @@
+
+As far as I've spoken to people, no one disagrees with the need to have some unit tests. That means we all agree that it's possible to write buggy code. That implies that our unit-tests could _also_ have bugs; how would we detect that if we wrote it second?
+
+## What do you mean by a buggy test, how can a test have a bug?
+
+The purpose of a test is to check the correctness of the function. So anything that prevents it from doing that, is a bug. i.e. if a test isn't actually checking anything then it's not doing it's job.
+
+Let's use the example above; let's say I finished implementing `mul2`, and then wrote the test below:
+
+```ts
+it('should return the product', () => {
+  const actual = mul2(2, 3)
+  const expected = 6
+})
+```
+
+Notice anything? The assertion statement is missing, but the actual test will pass[^false-positive]. Here even though everything looks okay, the test itself is not trustworthy.
+
+## What a ridiculous example, who would forget to add the assertion!
+
+Well yes it's a bit silly, but it highlights a central problem with unit testing. Namely:
+
+> How do you know that the test you've written is correct?
+
+Any alternative proposed to TDD **must** be able to solve this problem or else it cannot be said to be viable.
+
+The fact that the test is written first solves this problem. If the test is correct, it _must_ fail.
+
+In the example above you expect it to fail, so you'll revisit the test-code when you see it's passing even if no code is present. The rules push you to prioritize confidence in the test first before starting on the change you want to make.
+
+## Hmm... Well what about Logical Errors, TDD is not going to save me there
+
+TDD doesn't guarantee that your code is bug-free. It just guarantees that if a line is deleted that a test will fail. Therefore it's still your responsibility to write the correct tests. For example if the requirement was to multiply two numbers, but your tests are checking if the code adds the two together then your code is still wrong.
+
+In such a scenario, TDD helps in two ways:
+
+- it makes such an error easier to spot during code-review
+  - since your `expected` value will clearly be something different than the requirement, someone can flag it as an issue
+- once the error is spotted, it becomes easier to rewrite the bad test or to add tests for missing edge cases.
+
+Since refactoring is easier, this whole process happens much more quickly. And yes, while fixing one issue you're less likely to introduce a a regression even in this new feature.
+
+## This is all horribly complicated! I'm a pretty good programmer, I don't need all this
+
+I felt the same way when first starting out. Even now, when discussing some of the finer points of TDD it can feel a little overly-cautious and pessimistic. The thing is though, if there's more than 2 programmers working on a project it's just not possible to always be available for everyone.
```

Comment:
It took me a while to digest why the second phrase ("The thing is ...") is here. After re-reading, I see that it has to do with the second phrase in the next paragraph ("In the absence..."). May I suggest to get both statements closer/together?